addons/crm/models/crm_team_member.py

```diff
@@ -0,0 +1,188 @@
+              probability of being picked up next;
+          * if not found: consider this member is out of assignment process,
+            remove it from population so that it is not picked up anymore;
+
+        Assignment is performed one lead at a time for fairness purpose. Indeed
+        members may have overlapping domains within a given team. To ensure
+        some fairness in process once a member receives a lead, a new choice is
+        performed with updated weights. This is not optimal from performance
+        point of view but increases probability leads are correctly distributed
+        within the team.
+
+        :param int work_days: see ``CrmTeam.action_assign_leads()``;
+
+        :return members_data: dict() with each member assignment result:
+          membership: {
+            'assign': set of lead IDs directly assigned to the member;
+          }, ...
+
+        """
+        if not work_days or work_days > 30:
+            raise ValueError(
+                _('Leads assignments should be done for at least 1 or maximum 30 work days, not %s', work_days)
+            )
+        # assignment_max is valid for "30 days" -> divide by requested work_days
+        # to have number of leads to assign
+        assign_ratio = 30.0 / work_days
+
+        members_data, population, weights = dict(), list(), list()
+        members = self.filtered(lambda member: member.assignment_max > member.lead_month_count)
+        if not members:
+            return members_data
+
+        # prepare a global lead count based on total leads to assign to salespersons
+        lead_limit = sum(
+            min(
+                int(math.ceil(member.assignment_max / assign_ratio)),
+                (member.assignment_max - member.lead_month_count)
+            )
+            for member in members
+        )
+
+        # could probably be optimized
+        for member in members:
+            lead_domain = expression.AND([
+                literal_eval(member.assignment_domain or '[]'),
+                ['&', '&', ('user_id', '=', False), ('date_open', '=', False), ('team_id', '=', member.crm_team_id.id)]
+            ])
+
+            leads = self.env["crm.lead"].search(lead_domain, order='probability DESC', limit=lead_limit)
+
+            to_assign = min(member.assignment_max - member.lead_month_count, round(member.assignment_max / assign_ratio))
+            members_data[member.id] = {
+                "team_member": member,
+                "max": member.assignment_max,
+                "to_assign": to_assign,
+                "leads": leads,
+                "assigned": self.env["crm.lead"],
+            }
+            population.append(member.id)
+            weights.append(to_assign)
+
+        leads_done_ids = set()
+        counter = 0
+        while population and counter < 100:
+            counter += 1
+            member_id = random.choices(population, weights=weights, k=1)[0]
+            member_index = population.index(member_id)
```

Comment:
Feels a bit strange to me that we use such an advanced algorithm with randomness to assign leads to salesmen.

It means they're assigned leads based on rules they can't comprehend and they might feel very confused if "unluckily" they received 0 leads and their colleague received 5 (I get that the weight is there to counterbalance that, but it's still possible).

Would it be so bad to do:
David needs 5 leads
Thibault needs 2 leads
7 leads are available:
David gets 6
Thibault gets 1

I guess this is a simple use case but you get my meaning, remove the "randomness" of the process and make it strictly logical.
(And when you have to assign a lead between 2 persons and both have the same "available slots", then maybe use randomness at that point).

Side note: Or maybe this is to have an additional "AI" to talk about during OXP? Might be a valid argument.

Reply:
When weight reaches 0, member is removed from population. So David should never receives more than 5 leads. I suppose that due to members domains it is possible that "last" members do not receive all their leads (e.g. if David did receive leads matching other people domain). 

What is random is the ordering of assignment. It is weighted so that David has higher chances to receive first leads. It means that people with a high monthly volume get better leads than people with low volume ... maybe this is considered fair, or unfair, not sure.

Randomness is actually required in this process. It seems sales people want randomness as this is considered as more fair than some "logical" ordering ... But maybe I got your comment wrong.

Reply:
Yes, but the weight only goes down 1 by 1 right?

So if David has a capacity of 25, same as Thibault, and there are 10 leads to assign, it's still possible (even though very unlikely) that David receives all of the 10 leads and Thibault none?

I'm pushing the probabilities really far here to illustrate, of course actual unfairness cases should be pretty rare.
Although if the weights of all members start really high, and there is not many leads to assign, then it's very much possible, as the weight will not impact the probabilities a lot.

Random idea while I'm writing this.
To counterbalance overall high weights, has it been considered to divide the weight by 2 when you get a lead assigned (rounding upwards), instead of just subtracting 1?
Meaning that if you start really high weight, you will go down fast and allow other people to get leads, but if you start low on weight, you will not go down as fast.

There are just some wild thoughts while this is fresh in my head.
I would understand if you feel like the current mechanism is "good enough".